test/jdk/java/awt/dnd/DragSourceMotionListenerTest.java

```diff
@@ -0,0 +1,250 @@
+            if (Math.abs(dsde.getX() - testPoint2.getX()) < 5) {
+                passedTest2 = true;
+            }
+        }
+    };
+
+    private static final DropTargetListener targetAdapter = new DropTargetAdapter() {
+        public void drop(DropTargetDropEvent e) {
+            e.acceptDrop(DnDConstants.ACTION_COPY);
+            try {
+                final Transferable t = e.getTransferable();
+                final String str =
+                        (String) t.getTransferData(DataFlavor.stringFlavor);
+                e.dropComplete(true);
+            } catch (Exception ex) {
+                ex.printStackTrace();
+                e.dropComplete(false);
+            }
+        }
+    };
+
+    private static final DropTarget dropTarget = new DropTarget(target, targetAdapter);
+    Component clickedComponent = null;
+
+    private void createAndShowUI() {
+        frame = new Frame("DragSourceMotionListenerTest");
+        ds.addDragSourceListener(sourceAdapter);
+        ds.addDragSourceMotionListener(sourceAdapter);
+        ds.createDefaultDragGestureRecognizer(source, DnDConstants.ACTION_COPY, gestureListener);
+        target.setDropTarget(dropTarget);
+
+        frame.setLayout(new GridLayout(1, 2));
+
+        frame.add(source);
+        frame.add(target);
+
+        Toolkit.getDefaultToolkit()
+               .addAWTEventListener(this, AWTEvent.MOUSE_EVENT_MASK);
+        frame.pack();
+        frame.setVisible(true);
+    }
+
+    public static void main(String[] args) throws Exception {
+        final Dimension[] d = new Dimension[1];
+        try {
+            Robot robot = new Robot();
+            robot.setAutoDelay(10);
+
+            DragSourceMotionListenerTest dsmObj = new DragSourceMotionListenerTest();
+            EventQueue.invokeAndWait(dsmObj::createAndShowUI);
+            robot.waitForIdle();
+            robot.delay(1000);
+
+            EventQueue.invokeAndWait(() -> {
+                srcPoint = source.getLocationOnScreen();
+                d[0] = source.getSize();
+                srcPoint.translate(d[0].width / 2, d[0].height / 2);
+            });
+            robot.waitForIdle();
+
+            if (!dsmObj.pointInComponent(robot, srcPoint, source)) {
+                throw new RuntimeException("WARNING: Couldn't locate source panel.");
+            }
+
+            EventQueue.invokeAndWait(() -> {
+                dstOutsidePoint = frame.getLocationOnScreen();
+                d[0] = frame.getSize();
+                dstOutsidePoint.translate(3 * d[0].width / 2, d[0].height / 2);
+                testPoint1.setLocation(dstOutsidePoint);
+             });
+
+
+            EventQueue.invokeAndWait(() -> {
+                dstInsidePoint = target.getLocationOnScreen();
+                d[0] = target.getSize();
+                dstInsidePoint.translate(d[0].width / 2, d[0].height / 2);
+                testPoint2.setLocation(dstInsidePoint);
```

Comment:
The same issue:
```suggestion
                Point p = target.getLocationOnScreen();
                Dimension d = target.getSize();
                p.translate(d.width / 2, d.height / 2);
                dstInsidePoint = p;
                testPoint2.setLocation(p);
```

Reply:
In fact, you can merge the three calls to `invokeAndWait` into one which uses a private helper method.

The location and size of the components and the frame don't change after you first showed the UI.

So, you can calculate the value for `srcPoint`, `dstOutsidePoint` and `dstInsidePoint` in one call into `invokeAndWait`. It would look cleaner.

Reply:
Combined into one EDT call and helper method as below.
```
  EventQueue.invokeAndWait(() -> {
                Point p = getPoint(source, 1);
                srcPoint = p;

                p = getPoint(frame, 3);
                dstOutsidePoint = p;
                testPoint1.setLocation(p);

                p = getPoint(target, 1);
                dstInsidePoint = p;
                testPoint2.setLocation(p);
            });

//helper method

    private static Point getPoint(Container container, int multiple) {
        Point p = container.getLocationOnScreen();
        Dimension d = container.getSize();
        p.translate(multiple * d.width / 2, d.height / 2);
        return p;
    }
```